### PR TITLE
feat: add skywalking trace

### PR DESCRIPTION
**Pull request by stulzq**
**What this PR does**:
add skywalking tracer
**Which issue(s) this PR fixes**:

#310 
**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: diagnostics/grpc_tracing.go [hunk 1]
```diff
@@ -40,21 +47,23 @@ func StreamInterceptorFilter(srv interface{}, ss grpc.ServerStream, info *grpc.S
...
 	defer span.FinishSpan()
-	span.SetTag(ltrace.LAYOTTO_METHOD_NAME, info.FullMethod)
-	span.SetTag(ltrace.LAYOTTO_REQUEST_RESULT, "0")
+
 	// construct a new context which contains the span
 	wrapped := grpc_middleware.WrapServerStream(ss)
 	ctx = GetNewContext(ctx, span)
 	wrapped.WrappedContext = ctx
 	// handle request
-	err := handler(srv, wrapped)
```
Comment (wenxuwan) [inline diagnostics/grpc_tracing.go:55]: 这个地方改成handler(ctx, ss)，stream没法将ctx的header传递下去吧

File: diagnostics/grpc_tracing.go [hunk 2]
```diff
@@ -40,21 +47,23 @@ func StreamInterceptorFilter(srv interface{}, ss grpc.ServerStream, info *grpc.S
...
+	err := handler(ctx, wrapped)
```
Comment (wenxuwan) [inline diagnostics/grpc_tracing.go:63]: 第一个参数改成ctx的原因是什么，上下文应该在第二个参数里面的ctx中

Comment (stulzq) [reply diagnostics/grpc_tracing.go]: 我的锅 c/v 的时候没注意

**Review by wenxuwan — DISMISSED** [review]
LGTM

**Comment by seeflood** [comment]
Thanks !
I just changed the `config_stores` in config.json  to `config_store`

**Review by zhenjunMa — DISMISSED** [review]

**Comment by seeflood** [comment]
Got an error when running the demo:
image
It seems that the main.go forgot to import grpcmetric package. I will help fix it

**Comment by seeflood** [comment]
fixed. 
Besides, I modified the document and added it to the sidebar

File: diagnostics/grpc_tracing.go [hunk 3]
```diff
@@ -40,13 +47,14 @@ func StreamInterceptorFilter(srv interface{}, ss grpc.ServerStream, info *grpc.S
...
+	tracer := trace.Tracer(protocol.Layotto)
 	ctx := ss.Context()
 	// start a span
-	span := tracer.Start(ctx, nil, time.Now())
+	span := tracer.Start(ctx, &lgrpc.RequestInfo{
+		FullMethod: info.FullMethod,
+	}, time.Now())
 	defer span.FinishSpan()
-	span.SetTag(ltrace.LAYOTTO_METHOD_NAME, info.FullMethod)
-	span.SetTag(ltrace.LAYOTTO_REQUEST_RESULT, "0")
```
Comment (seeflood) [inline diagnostics/grpc_tracing.go:49]: Why delete them?

File: diagnostics/skywalking/grpc_tracer.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+		ctx:    nCtx,
+		carrier: &skywalking.SpanCarrier{
+			EntrySpan: entry,
+		},
+		Span: &ltrace.Span{},
+	}
+}
+
+type grpcSkySpan struct {
+	*ltrace.Span
```
Comment (seeflood) [inline diagnostics/skywalking/grpc_tracer.go:83]: So `grpcSkySpan.SetTag` will actually invoke `*ltrace.Span.SetTag` .
But the tags will be ignored when `grpcSkySpan.End()` . So finally we lost the tags.
Am I right? I think these tags should be sent to SkyWalking too.

**Review by seeflood — CHANGES_REQUESTED** [review]

File: diagnostics/grpc_tracing.go (hunk 3 above)
Comment (stulzq) [reply diagnostics/grpc_tracing.go:49]: 移动到了这边 diagnostics/tracing.go

Comment (stulzq) [reply diagnostics/grpc_tracing.go:49]: https://github.com/mosn/layotto/pull/404/files#diff-360719837a63f9a6d3b8c94a5bc6f97daa5901a20bd04a620e2298aced617032R51

File: diagnostics/skywalking/grpc_tracer.go (hunk 4 above)
Comment (stulzq) [reply diagnostics/skywalking/grpc_tracer.go:83]: sky 不支持自定义 tag , https://github.com/mosn/mosn/blob/master/pkg/trace/skywalking/span.go#L53

Comment (seeflood) [reply diagnostics/skywalking/grpc_tracer.go:83]: 好吧，有点奇怪……那先记个TODO 以后再优化吧

File: diagnostics/grpc_tracing.go (hunk 3 above)
Comment (seeflood) [reply diagnostics/grpc_tracing.go:49]: 恩恩看到了，主要是感觉，更适合把SetTag放在这里，因为：
1. `UnaryInterceptorFilter`是放这的，如果`StreamInterceptorFilter`不放这的话，不一致（强迫症看着难受 
image

2. 放这的好处是不用每个span都去写这两行了（虽然现在skywalking的span 没法setTag）
3. 上下一致，下面有个set "1", 如果把上面的 set "0"删了 看着有点不一致
image

Comment (stulzq) [reply diagnostics/grpc_tracing.go:49]: 我改下，当时移动的初衷是不是每个span
都需要设置这个

File: docs/zh/start/trace/skywalking.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
+# Skywalking trace 接入
```
Comment (seeflood) [inline docs/zh/start/trace/skywalking.md:1]: Could u also add this great document into our README-zh.md?
https://mosn.io/layotto/#/zh/README

I think we can add a new table for Observability
image

**Review by seeflood — CHANGES_REQUESTED** [review]

**Review by seeflood — APPROVED** [review]
lgtm. Thanks for your great contribution!

**Review by zhenjunMa — APPROVED** [review]

**Pull request merged**